File: 15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        for (int a = 0; a < n; a++)
        {
            // skip continues same number
            if (a > 0 && nums[a] == nums[a - 1])
            {
                continue;
            }
            // after c
            int c = n - 1;
            int target = -nums[a];
            for (int b = a + 1; b < n; b++)
            {
                if (b > a + 1 && nums[b] == nums[b - 1])
                {
                    continue;
                }
                while (b < c && nums[b] + nums[c] > target)
                {
                    --c;
                }
                if (b == c)
                {
                    break;
                }
                if (nums[b] + nums[c] == target)
                {
                    ans.push_back({nums[a], nums[b], nums[c]});
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: JavaScript/15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  nums.sort((a, b) => a - b);

  const result = [];
  const n = nums.length;

  for (let a = 0; a < n - 2; a++) {
    if (a > 0 && nums[a] === nums[a - 1]) {
      continue;
    }

    let b = a + 1;
    let c = n - 1;
    let target = -nums[a];

    while (b < c) {
      const sum = nums[b] + nums[c];

      if (sum === target) {
        result.push([nums[a], nums[b], nums[c]]);
        while (b < c && nums[b] === nums[b + 1]) b++;
        while (b < c && nums[c] === nums[c - 1]) c--;
        b++;
        c--;
      } else if (sum < target) {
        b++;
      } else {
        c--;
      }
    }
  }

  return result;
};
// @lc code=end
